### PR TITLE
Lighting/urp reflectionprobe tests2

### PR DESCRIPTION
**Pull request by pigselated**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

---
### Purpose of this PR
This PR is adding automated tests to Reflection Probe baking in URP. It is in response to recent HUP bugs related to that area. Refer to this RCA for details:
https://confluence.unity3d.com/display/teamgi/Reflection+Probe+Baking+Issues

---
### Testing status
This PR adds only automated test and no new functionality.

---
### Comments to reviewers
The editor that this test works correctly against is this:
https://ono.unity3d.com/unity/unity/pull-request/124303/_/lighting/case1324246-missing-indirect-in-ref-probe

### CODE REVIEW

**Review by KEngelstoft — CHANGES_REQUESTED** [review]
Formatting job is failing:
Illegal characters or use of reserved words in path/filename: TestProjects/UniversalGraphicsTest_Lighting/Assets/ReferenceImages/Linear/WindowsEditor/Direct3D11/None/RefProbeAPI(nonAuto-nonProgressive,BakeSingle).png.meta (see http://go/ihudr)

**Comment by KEngelstoft** [comment]
There are diffs in unrelated files, like 005 and 151. Please revert unrelated changes.
Example https://github.com/Unity-Technologies/Graphics/blob/2befbbbb9ea91f0df2b384d9ebcabb46ed8c0554/TestProjects/UniversalGraphicsTest_Lighting/Assets/ReferenceImages/Linear/LinuxEditor/OpenGLCore/None/005_LitBakedEmission.png

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/CopyImageToReferenceFolders.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
+using UnityEngine;
+using UnityEditor;
+using System.IO;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+
+public class CopyImageToReferenceFolders
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/CopyImageToReferenceFolders.cs:8]: Why duplicate the file? It is already available in com.unity.testing.urp/Scripts/Editor/CopyImageToReferenceFolders.cs

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+using NUnit.Framework;
+using UnityEditor;
+using UnityEngine;
+using UnityEngine.Profiling;
+
+[TestFixture]
+public class MultipleViewGCTest : MonoBehaviour
+{
+    Recorder m_gcAllocRecorder;
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs:9]: Why is MultipleViewGCTest needed?

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+using UnityEngine;
+using NUnit.Framework;
+using UnityEngine.TestTools.Graphics;
+using UnityEditor.SceneManagement;
+using UnityEditor;
+
+public class Editmode_ParametricReflectionProbeTests
+{
+	private readonly string sceneOutputPath = "Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/ReflectionProbeBake";
+	private readonly string sceneFileName = "Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/ReflectionProbeBake.unity";
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs:12]: Looks like you don't have the formatting githook installed? https://github.com/Unity-Technologies/Graphics/tree/master/Tools

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/UniversalProjectEditorTests.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    private static UniversalRenderPipelineAsset currentAsset;
+
+    [Test]
+    public void GetCurrentAsset()
+    {
+        GetUniversalAsset();
+    }
+
+    [Test]
+    public void CheckAllLightingSettings()
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/UniversalProjectEditorTests.cs:18]: Why duplicate? It is available here com.unity.testing.urp/Scripts/Editor/UniversalProjectEditorTests.cs

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Runtime/ArrayDrawer.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
+using UnityEngine;
+
+[ExecuteInEditMode]
+public class ArrayDrawer : MonoBehaviour
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Runtime/ArrayDrawer.cs:4]: why duplicate, it is available here com.unity.testing.urp/Scripts/Runtime/ArrayDrawer.cs

**Comment by KEngelstoft** [comment]
Please let me know when this PR is ready for another review iteration.

**Comment by pigselated** [comment]
> Please let me know when this PR is ready for another review iteration.

Many of the issues you mention are doing to @phi-lira 's recent changes and the revert of them:
https://github.com/Unity-Technologies/Graphics/pull/4257/commits/4fa77d70b2f59c9e30c9fdd06ead62196b2597fb

I simply merged his reverted changes into my branch. When I merged master the tests broke, and @phi-lira reverted the changes and I merge that in to undo the changes that came with master. Please disregard these changes.

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs (hunk 2 above)
Comment (pigselated) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs:9]: Same as above...

This was some refactoring done by URP team and then it was reverted. The revert PR is merged to my branch to make things work again:
https://github.com/Unity-Technologies/Graphics/pull/4257/commits/5b8fb9220633fffdfcf9189fbec2eb645fe6508c

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/UniversalProjectEditorTests.cs (hunk 4 above)
Comment (pigselated) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/UniversalProjectEditorTests.cs:18]: This was some refactoring done by URP team and then it was reverted. The revert PR is merged to my branch to make things work again:
https://github.com/Unity-Technologies/Graphics/pull/4257/commits/5b8fb9220633fffdfcf9189fbec2eb645fe6508c

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Runtime/ArrayDrawer.cs (hunk 5 above)
Comment (pigselated) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Runtime/ArrayDrawer.cs:4]: This was some refactoring done by URP team and then it was reverted. The revert PR is merged to my branch to make things work again:
https://github.com/Unity-Technologies/Graphics/pull/4257/commits/5b8fb9220633fffdfcf9189fbec2eb645fe6508c

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs (hunk 3 above)
Comment (pigselated) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs:12]: Will install them now. Looks like the installation doesn't work with the new python versions. Will check with @theo-pnv .

**Review by phi-lira — APPROVED** [review]

**Comment by pigselated** [comment]
> Formatting job is failing:
> Illegal characters or use of reserved words in path/filename: TestProjects/UniversalGraphicsTest_Lighting/Assets/ReferenceImages/Linear/WindowsEditor/Direct3D11/None/RefProbeAPI(nonAuto-nonProgressive,BakeSingle).png.meta (see http://go/ihudr)

Do you think it is the ","? I have removed the commas in the filenames now and have replaced them with "_" instead.

**Comment by theo-pnv** [comment]
For these files `TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/red emissive.mat`, it doesn't like the spaces.
For these `TestProjects/UniversalGraphicsTest_Lighting/Assets/ReferenceImages/Linear/WindowsEditor/Direct3D11/None/RefProbeAPI(nonAuto-nonProgressive_Bake).png`, my guess is it doesn't like the parenthesis (can't see anything else that would cause issues from experience)

**Review by KEngelstoft — CHANGES_REQUESTED** [review]
Unrelated test images (005, 152 and 153) are still there.

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs (hunk 2 above)
Comment (KEngelstoft) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/MultipleViewGCTest.cs:9]: Please clarify, will a future merge of master make these files disappear from the diff?

**Comment by pigselated** [comment]
> Unrelated test images (005, 152 and 153) are still there.

@KEngelstoft these came from the merge of @phi-lira's branch. Basically a revert of some changes made to master that broke the tests in this PR.

File: .yamato/config/universal_lighting.metafile [hunk 6]
```diff
@@ -127,9 +127,10 @@ platforms:
         flavor: b1.xlarge
         model: rtx2080
       editmode:
-        type: Unity::VM
+        type: Unity::VM::GPU
         image: sdet/gamecode_win10:stable
-        flavor: b1.large
+        flavor: b1.xlarge
+        model: rtx2080
```
Comment (theo-pnv) [inline .yamato/config/universal_lighting.metafile:133]: The xlarge flavor is unfortunately not available for GPUs, you will have to use the b1.large one https://internaldocs.hq.unity3d.com/bokken/reference/flavors/

**Review by theo-pnv — CHANGES_REQUESTED** [review]
Please remove the spaces from the following file names, as we won't be able to mirror the graphics code to unity code if the formatting job does not pass here
```
TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/green emissive.mat.meta
TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/red emissive.mat
TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/green emissive.mat TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/red emissive.mat.meta 
```

File: .yamato/config/universal_lighting.metafile (hunk 6 above)
Comment (pigselated) [reply .yamato/config/universal_lighting.metafile:133]: Ok will do

**Comment by pigselated** [comment]
> Please remove the spaces from the following file names, as we won't be able to mirror the graphics code to unity code if the formatting job does not pass here
> 
> ```
> TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/green emissive.mat.meta
> TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/red emissive.mat
> TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/green emissive.mat
> TestProjects/UniversalGraphicsTest_Lighting/Assets/EditModeTestAssets/Lighting_ReflectionProbeBaking/contents/red emissive.mat.meta 
> ```

Fixed this.

**Comment by pigselated** [comment]
> Unrelated test images (005, 152 and 153) are still there.

This has been fixed.

File: .yamato/urp_lighting-win-dx11.yml [hunk 7]
```diff
@@ -140,9 +140,10 @@ URP_Lighting_Win_DX11_playmode_cache_mono_Linear_trunk:
 URP_Lighting_Win_DX11_editmode_mono_Linear_trunk:
     name: URP_Lighting on Win_DX11_editmode_mono_Linear on version trunk
     agent:
-        type: Unity::VM
+        type: Unity::VM::GPU
         image: sdet/gamecode_win10:stable
-        flavor: b1.large
+        flavor: b1.xlarge
```
Comment (KEngelstoft) [inline .yamato/urp_lighting-win-dx11.yml:145]: @sophiaaar FYI ^

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+		Debug.Log("Found " + settingsAssets.Length + " matching lighting settings assets in " + foldersToLookIn[0]);
+		Assert.That(settingsAssets.Length, Is.EqualTo(1));
+		string lsaPath = AssetDatabase.GUIDToAssetPath(settingsAssets[0]);
+		Debug.Log("Loading " + lsaPath);
+		LightingSettings lightingSettings = (LightingSettings)AssetDatabase.LoadAssetAtPath(lsaPath, typeof(LightingSettings));
+		Lightmapping.lightingSettings = lightingSettings;
+		string fileName = System.IO.Path.GetFileNameWithoutExtension(lsaPath);
+		Assert.That(fileName, Is.EqualTo(settings));
+		Lightmapping.Clear();
+        Lightmapping.ClearDiskCache();
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs:79]: please add a comment why you clear

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    [TestCaseSource("GetReflectionProbeTestCases")]
+    public void RefProbeAPI(string settings, BakeAPI bakeAPI)
+	{
+		EditorSceneManager.OpenScene(sceneFileName, OpenSceneMode.Single);
+
+		// Bake with a lighting settings asset.
+		string[] settingsAssets = AssetDatabase.FindAssets(settings + " t:lightingsettings", foldersToLookIn);
+		Debug.Log("Found " + settingsAssets.Length + " matching lighting settings assets in " + foldersToLookIn[0]);
+		Assert.That(settingsAssets.Length, Is.EqualTo(1));
```
Comment (KEngelstoft) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs:71]: formatting?

**Review by KEngelstoft — CHANGES_REQUESTED** [review]
please revert the lock file

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs (hunk 8 above)
Comment (pigselated) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Test/Editor/ReflectionProbeTests.cs:82]: Fixed

File: .yamato/urp_lighting-win-dx11.yml (hunk 7 above)
Comment (pigselated) [reply .yamato/urp_lighting-win-dx11.yml:145]: Changed it back to b1.large as requested by @theo-pnv

**Review by KEngelstoft — APPROVED** [review]

File: .yamato/config/universal_lighting.metafile [hunk 10]
```diff
@@ -127,9 +127,10 @@ platforms:
         flavor: b1.xlarge
         model: rtx2080
       editmode:
```
Comment (KEngelstoft) [inline .yamato/config/universal_lighting.metafile:129]: Why are all the yml changes also present in the meta file? Not specific to this PR...

File: .yamato/urp_lighting-win-dx11.yml (hunk 7 above)
Comment (theo-pnv) [reply .yamato/urp_lighting-win-dx11.yml]: For playmode and standalone jobs the GPU is needed (with `b1.large` flavor) but for editmode we don't need the GPU though. So you can keep `Unity::VM` and `b1.xlarge` for those.

**Review by theo-pnv — CHANGES_REQUESTED** [review]

File: .yamato/urp_lighting-win-dx11.yml (hunk 7 above)
Comment (KEngelstoft) [reply .yamato/urp_lighting-win-dx11.yml]: The new reflection probe tests are edit mode, so they need a GPU right?

Comment (pigselated) [reply .yamato/urp_lighting-win-dx11.yml]: Yes, thats true. I also want to add GPU LM tests to that suite.

File: .yamato/config/universal_lighting.metafile (hunk 10 above)
Comment (sophiaaar) [reply .yamato/config/universal_lighting.metafile:129]: there are specific changes for this PR - this is the first time we're using a GPU for editmode tests

File: .yamato/urp_lighting-win-dx12.yml [hunk 11]
```diff
@@ -140,9 +140,10 @@ URP_Lighting_Win_DX12_playmode_cache_mono_Linear_trunk:
 URP_Lighting_Win_DX12_editmode_mono_Linear_trunk:
     name: URP_Lighting on Win_DX12_editmode_mono_Linear on version trunk
     agent:
-        type: Unity::VM
+        type: Unity::VM::GPU
         image: sdet/gamecode_win10:stable
         flavor: b1.large
```
Comment (sophiaaar) [inline .yamato/urp_lighting-win-dx12.yml:145]: you need to use b1.xlarge to be able to use the rtx2080s, can you please revert these changes? thanks

**Review by sophiaaar — CHANGES_REQUESTED** [review]

File: .yamato/config/universal_lighting.metafile (hunk 10 above)
Comment (pigselated) [reply .yamato/config/universal_lighting.metafile:129]: @sophiaaar still what is the purpose of the metafile?

File: .yamato/urp_lighting-win-dx12.yml (hunk 11 above)
Comment (pigselated) [reply .yamato/urp_lighting-win-dx12.yml:145]: will do

Comment (pigselated) [reply .yamato/urp_lighting-win-dx12.yml]: done in 264ff4ed6d52ddf676be573920d2311e81bf5627

File: .yamato/config/universal_lighting.metafile (hunk 10 above)
Comment (pigselated) [reply .yamato/config/universal_lighting.metafile:129]: does it also need to be set to b1.xlarge?

**Comment by pigselated** [comment]
> please revert the lock file

Done

**Comment by pigselated** [comment]
@sophiaaar @theo-pnv I think I have fixed all the requested issues. Can you have a look?

File: .yamato/config/universal_lighting.metafile (hunk 10 above)
Comment (sophiaaar) [reply .yamato/config/universal_lighting.metafile:129]: the metafile is what we use to define the rules for the jobs, and then we run a script to generate the ymls. this is so we avoid things like copy-paste errors etc

Comment (sophiaaar) [reply .yamato/config/universal_lighting.metafile:129]: can i make a commit on your branch just to check the ymls have all been generated?

**Comment by pigselated** [comment]
Hmm, formatting job has just failed. There are no actionable errors in the log:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/lighting%252FURP-reflectionprobe-tests2/.yamato%252F_formatting.yml%2523formatting/6698710/job

There's a couple of warnings on unrelated files.

Log ends with this:

[07:31:54.361 Information] Process exited
[07:31:54.367 Information] Command failed.
[07:31:54.391 Information] Unexpected ExitCode 1 (expected 0)

Is this normal @theo-pnv ?

**Comment by theo-pnv** [comment]
@pigselated I think merging master into this branch will solve the formatting issues. I'm not really sure why these files show up

**Review by theo-pnv — APPROVED** [review]

**Review by sophiaaar — APPROVED** [review]

**Review by sophiaaar — APPROVED** [review]

**Comment by phi-lira** [comment]
Yamato failures not related to this PR.
